pennylane/qchem/givens_decomposition.py

```diff
@@ -38,26 +36,50 @@ def _givens_matrix(a, b, left=True, tol=1e-8):
         tol (float): determines tolerance limits for :math:`|a|` and :math:`|b|` under which they are considered as zero.
 
     Returns:
-        np.ndarray (or tensor): Givens rotation matrix
+        tensor_like: Givens rotation matrix
 
     """
-    abs_a, abs_b = np.abs(a), np.abs(b)
-    if abs_a < tol:
-        cosine, sine, phase = 1.0, 0.0, 1.0
-    elif abs_b < tol:
-        cosine, sine, phase = 0.0, 1.0, 1.0
-    else:
-        hypot = np.hypot(abs_a, abs_b)
-        cosine = abs_b / hypot
-        sine = abs_a / hypot
-        phase = 1.0 * b / abs_b * a.conjugate() / abs_a
+    abs_a, abs_b, interface = qml.math.abs(a), qml.math.abs(b), qml.math.get_interface(a)
+    aprod = qml.math.nan_to_num(abs_b * abs_a)
+    hypot = qml.math.hypot(abs_a, abs_b)
+
+    cosine = qml.math.where(abs_b < tol, 0.0, abs_b / hypot)
+    cosine = qml.math.where(abs_a < tol, 1.0, cosine)
+
+    sine = qml.math.where(abs_b < tol, 1.0, abs_a / hypot)
+    sine = qml.math.where(abs_a < tol, 0.0, sine)
+
+    phase = qml.math.where(abs_b < tol, 1.0, (1.0 * b * qml.math.conj(a)) / (aprod + 1e-15))
+    phase = qml.math.where(abs_a < tol, 1.0, phase)
 
     if left:
-        return np.array([[phase * cosine, -sine], [phase * sine, cosine]])
+        return qml.math.array([[phase * cosine, -sine], [phase * sine, cosine]], like=interface)
+
+    return qml.math.array([[phase * sine, cosine], [-phase * cosine, sine]], like=interface)
+
+
+def _set_unitary_matrix(unitary_matrix, index, value, like=None):
+    """Set the values in the ``unitary_matrix`` at the specified index.
+
+    Args:
+        unitary_matrix (tensor_like): unitary being modified
+        index (Tuple[int | Ellipsis | List[Int]]): index for the slicing the unitary
```

Comment:
Yes, the algorithm must build the left and right decompositions more efficiently via vectorization, for which this is required. Since the presence of `Ellipsis` allows for obtaining/setting a sequence rather than a specific element, I think `slicing` is more appropriate.